### PR TITLE
Require fixed width types for casting in `cudf-polars`

### PR DESCRIPTION
**Pull request by brandon-b-miller**
Fixes a bug where numeric <-> string casts are not being properly rejected at the cudf-polars level.

### CODE REVIEW

**Review by vyasr — APPROVED** [review]

File: python/cudf_polars/cudf_polars/dsl/expr.py [hunk 1]
```diff
@@ -1188,7 +1188,11 @@ class Cast(Expr):
     def __init__(self, dtype: plc.DataType, value: Expr) -> None:
         super().__init__(dtype)
         self.children = (value,)
-        if not plc.unary.is_supported_cast(self.dtype, value.dtype):
+        if (
+            not plc.traits.is_fixed_width(self.dtype)
+            and plc.traits.is_fixed_width(value.dtype)
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/dsl/expr.py:1193]: This logic looks wrong to me. 

The supported casts are: `is_fixed_width(from) and is_fixed_width(to) and is_supported_cast(from, to)`, so you want to raise for the converse

```
if not (plc.traits.is_fixed_width(self.dtype) and plc.traits.is_fixed_width(value.dtype) and plc.unary.is_supported_cast(value.dtype, self.dtype):
    raise ...
```

Note also the change of order in `is_supported_cast`. `value` provides the source dtype and `self` provides the value dtype.

**Review by wence- — CHANGES_REQUESTED** [review]
I think the logic is not right, sorry

File: python/cudf_polars/cudf_polars/dsl/expr.py (hunk 1 above)
Comment (vyasr) [reply python/cudf_polars/cudf_polars/dsl/expr.py:1193]: Hmm perhaps I misunderstood, I thought this was allowing a cast from a non-fixed to fixed. Do we need both types to be fixed width?

Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/expr.py:1193]: `cudf::cast` docstring says ` * Supports only fixed-width types.` But `is_supported_cast` only answers correctly if the inputs are fixed-width.

Comment (wence-) [reply python/cudf_polars/cudf_polars/dsl/expr.py:1193]: I didn't realise this when I introduced the `cudf::is_supported_cast(from_dtype, to_dtype)` public function.

**Comment by vyasr** [comment]
@brandon-b-miller it looks like the new code path in this test is not fully being exercised. If I had to guess, it's because the condition is short-circuiting so some of the later conditions are never being evaluated (i.e. you don't reach `is_supported_cast` if both `is_fixed_width` tests fail).

**Comment by brandon-b-miller** [comment]
> @brandon-b-miller it looks like the new code path in this test is not fully being exercised. If I had to guess, it's because the condition is short-circuiting so some of the later conditions are never being evaluated (i.e. you don't reach `is_supported_cast` if both `is_fixed_width` tests fail).

This was subtle but it came down to a failed `StringFunction.Contains` test:
- the plan inserting a cast to string on the argument passed to contains (in this case an int->string cast)
- the existing code failing to reject this cast when setting up the test data and allowing the test to proceed and hit the NotImplementedError in contains that coverage now complains we miss
- The code still raising NotImplementedError eventually causing the test to pass both ways

**Comment by vyasr** [comment]
Nice find! That seems to have done the trick.

**Comment by vyasr** [comment]
I'll leave this for @wence- to re-review and merge in the morning, but hopefully this is all set now.

**Comment by brandon-b-miller** [comment]
> /merge

I think you'll need to dismiss your change/approve

**Pull request merged**